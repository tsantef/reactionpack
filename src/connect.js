import _ from 'lodash';
import React from 'react';

import { bindActions } from './actions';
import { buildComputeds, generateComputeds } from './computed';

function composeState({ propTypeKeys, defaults, containerState, boundComputeds, props }) {
	const __propKeys = _.keys(props);
	const __stateKeys = _.without(propTypeKeys, ...__propKeys);
	const composedState = {
		__propKeys,
		__stateKeys,
		__computed: null,
	};
	if (Object.keys(boundComputeds).length > 0) {
		const passThroughProps = _.pick(props, propTypeKeys);
		const stateToProps = _.pick(containerState, __stateKeys);
		const state = {
			...defaults,
			...stateToProps,
			...passThroughProps,
		};
		composedState.__computed = generateComputeds(boundComputeds, state);
	}
	return composedState;
}

export function connectToProps(WrappedComponent, actions, computeds, nameSpace) {
	const boundComputeds = buildComputeds(computeds);

	return React.createClass({
		propTypes: {
			statePath: React.PropTypes.string,
		},

		getDefaultProps() {
			return {
				statePath: '',
			};
		},

		getChildContext: function() {
			return {
				parentPath: this.getStatePath(),
			};
		},

		childContextTypes: {
			parentPath: React.PropTypes.array,
		},

		contextTypes: {
			parentPath: React.PropTypes.array,
			getState: React.PropTypes.func,
			setState: React.PropTypes.func,
		},

		getStatePath() {
			return _.compact((this.context.parentPath || []).concat([nameSpace, this.props.statePath]));
		},

		getInitialState() {
			const {
				statePath,
				...props
			} = this.props;
			const __propKeys = _.keys(props);
			const __propTypeKeys = _.keys(WrappedComponent.propTypes);
			const __actionKeys = _.keys(actions);
			const __computedKeys = _.keys(boundComputeds);
			const __stateKeys = _.without(__propTypeKeys, ...__propKeys);
			const getDefaultProps = WrappedComponent.getDefaultProps
				? WrappedComponent.getDefaultProps
				: () => (WrappedComponent.defaultProps || {});

			return {
				__nameSpace: nameSpace,
				__statePath: statePath,
				__propKeys,
				__propTypeKeys,
				__actionKeys,
				__computedKeys,
				__stateKeys,
				__actions: bindActions(this, actions),
				__computed: null,
				__defaults: getDefaultProps(),
			};
		},

		componentWillMount() {
			const {
				__defaults,
				__propTypeKeys,
			} = this.state;
			this.setState(composeState({
				defaults: __defaults,
				propTypeKeys: __propTypeKeys,
				containerState: this.context.getState(this.getStatePath()),
				props: this.props,
				boundComputeds,
			}));
		},

		componentWillReceiveProps(nextProps) {
			const {
				__defaults,
				__propTypeKeys,
			} = this.state;
			this.setState(composeState({
				defaults: __defaults,
				propTypeKeys: __propTypeKeys,
				containerState: this.context.getState(this.getStatePath()),
				props: nextProps,
				boundComputeds,
			}));
		},

		render() {
			const {
				__propTypeKeys,
				__stateKeys,
				__actions,
				__computed,
				__defaults,
			} = this.state;

			const passThroughProps = _.pick(this.props, __propTypeKeys);
			const stateToProps = _.pick(this.context.getState(this.getStatePath()), __stateKeys);
			// TODO: should cache the merge smartly
			const props = {..._.merge({}, __defaults, stateToProps, __actions, __computed, passThroughProps) };

			return React.createElement(WrappedComponent, props);
		},
	});
}
