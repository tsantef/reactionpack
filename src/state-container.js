import _ from 'lodash';
import React from 'react';

export function createStateContainer(WrappedComponent, initialState={}) {
	return React.createClass({
		propTypes: {
			onNextState: React.PropTypes.func,
			installDevTools: React.PropTypes.func, // Use for debugging purposes only
		},

		getChildContext() {
			return {
				getState: this._getState,
				setState: this._setState,
			};
		},

		childContextTypes: {
			getState: React.PropTypes.func,
			setState: React.PropTypes.func,
		},

		getInitialState() {
			return initialState;
		},

		_getState(path) {
			const compactPath = _.compact(path);
			if (!_.isEmpty(compactPath)) {
				return _.get(this.state, compactPath);
			} else {
				return this.state;
			}
		},

		_setState(path, state, actionName) {
			return new Promise((resolve) => {
				let newState;
				const compactPath = _.compact(path);
				if (!_.isEmpty(compactPath)) {
					newState = _.set({
						...this.state,
					}, compactPath, {
						..._.get(this.state, compactPath),
						...state,
					});
				} else {
					newState = {
						...this.state,
						...state,
					};
				}
				if (this.props && this.props.onNextState) {
					this.props.onNextState(newState, actionName);
				}
				if (this.onNextState) {
					this.onNextState(newState, actionName);
				}
				this.setState(newState, () => {
					resolve(this.state);
				});
			});
		},

		componentDidMount() {
			if (this.props.installDevTools) {
				// Call installDevTools with callback to reset state (development purposes only)
				this.props.installDevTools(
					(onNextState) => {
						this.onNextState = onNextState;
					},
					(resetState) => {
						this.setState(resetState);
					},
					this._getState()
				);
			}
		},

		render() {
			const props = _.omit(this.props, 'onNextState');
			return React.createElement(WrappedComponent, {...this.state, ...props});
		},
	});
}
