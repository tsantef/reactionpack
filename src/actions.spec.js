import _ from 'lodash';
import React from 'react';
import renderer from 'react-test-renderer';

import {
	bindActions,
	mockActions,
} from './actions';
import { connectToProps } from './connect';
import { createStateContainer } from './state-container';

describe('Actions', () => {

	describe('#bindActions', () => {

		it('root function', () => {

			const _this = {
				context: {
					getState: jest.fn(),
					setState: jest.fn(),
				},
				state: {
					__defaults: {},
					__propTypeKeys: ['value1'],
					__actionKeys: [],
					__computedKeys: [],
				},
				props: {
					value1: 4,
				},
			};

			const boundActions = bindActions(_this, {
				myAction: (state, value) => {
					expect(state).toEqual(_this.props);
					expect(value).toEqual(5);
					return { myState: 6 };
				},
			});
			boundActions.myAction(5);

			expect(_this.context.setState).toHaveBeenCalledWith([undefined, undefined, undefined, undefined], { myState: 6 }, 'myAction');

		});

		describe('context', () => {

			describe('#getState', () => {

				it('should existing in action context', () => {
					const _this = {
						state: {
							__defaults: {},
							__propTypeKeys: [],
							__actionKeys: [],
							__computedKeys: [],
						},
						context: {
							getState: jest.fn(),
							setState: jest.fn(),
						},
					};

					const boundActions = bindActions(_this, {
						myAction: function(state, value) {
							expect(this.getState).toBeDefined();
						},
					});

					boundActions.myAction();
				});

			});

			describe('#getComputed', () => {

				it('should existing in action context', () => {
					const _this = {
						context: {
							getState: jest.fn(),
							setState: jest.fn(),
						},
						state: {
							__propTypeKeys: [],
							__actionKeys: [],
							__computedKeys: [],
							__defaults: {
								value1: 'd1',
							},
							__computed: {
								value1: 'c1',
							},
						},
					};

					const boundActions = bindActions(_this, {
						myAction: function(state, value) {
							expect(this.getComputed).toBeDefined();
							expect(this.getComputed()).toEqual({ value1: 'c1' });
						},
					});

					boundActions.myAction();

				});

			});

			describe('#getActions', () => {

				it('should existing in action context', () => {

					const _this = {
						state: {
							__propTypeKeys: [],
							__actionKeys: [],
							__computedKeys: [],
						},
						context: {
							getState: jest.fn(),
							setState: jest.fn(),
						},
					};

					const boundActions = bindActions(_this, {
						myAction: function(state, value) {
							expect(this.getActions).toBeDefined();
							expect(this.getActions().myAction.name).toEqual('bound wrappedAction');
						},
					});

					boundActions.myAction();

				});

			});

			it('#getDefaults', () => {

				const _this = {
					context: {
						getState: jest.fn(),
						setState: jest.fn(),
					},
					state: {
						__propTypeKeys: [],
						__actionKeys: [],
						__computedKeys: [],
						__defaults: {
							value1: 'd1',
						},
						__computed: {
							value1: 'c1',
						},
					},
				};

				const boundActions = bindActions(_this, {
					myAction: function(state, value) {
						expect(this.getDefaults).toBeDefined();
						expect(this.getDefaults()).toEqual({ value1: 'd1' });
					},
				});

				boundActions.myAction();

			});

		});

		it('nested function', () => {

			const _this = {
				state: {
					__propKeys: [],
					__propTypeKeys: ['value2'],
					__actionKeys: [],
					__computedKeys: [],
				},
				context: {
					getState: jest.fn(),
					setState: jest.fn(),
				},
				props: {
					value1: 4,
					nest: {
						value2: 3,
					},
				},
			};

			const boundActions = bindActions(_this, {
				nest: {
					myAction: (state, value) => {
						expect(state).toEqual(_this.props.nest);
						expect(value).toEqual(5);
						return { myState: 6 };
					},
				},
			});

			boundActions.nest.myAction(5);

			expect(_this.context.setState).toHaveBeenCalledWith([undefined, undefined, undefined, 'nest'], { myState: 6 }, 'myAction');

		});

	});

	describe('wrappedAction', () => {

		it('should resolve return object', () => {

			const actions = {
				action1: () => {
					return { prop1: 'action1' };
				},
			};

			const Component = (props) => {
				return (
					<div {...props}></div>
				);
			};

			Component.propTypes = {
				prop1: React.PropTypes.string,
			};

			const BoundComponent = connectToProps(Component, actions);

			const StateContainer = createStateContainer(BoundComponent);

			const component = renderer.create(
				<StateContainer></StateContainer>
			);

			let tree = component.toJSON();
			let props = tree.props;

			props.action1();

			tree = component.toJSON();
			props = tree.props;

			expect(_.omit(props, 'action1')).toEqual({ prop1: 'action1' });

		});

		it('should allow null', () => {

			const actions = {
				action1: () => {
					return null;
				},
			};

			const Component = (props) => {
				return (
					<div {...props}></div>
				);
			};

			Component.propTypes = {
				prop1: React.PropTypes.string,
			};

			const BoundComponent = connectToProps(Component, actions);

			const StateContainer = createStateContainer(BoundComponent);

			const component = renderer.create(
				<StateContainer></StateContainer>
			);

			let tree = component.toJSON();
			let props = tree.props;

			props.action1();

			tree = component.toJSON();
			props = tree.props;

			expect(_.omit(props, 'action1')).toEqual({});

		});

		it('should reject invalid action', () => {

			const actions = {
				action1: 5,
			};

			const Component = (props) => {
				return (
					<div {...props}></div>
				);
			};

			Component.propTypes = {
				prop1: React.PropTypes.string,
			};

			const BoundComponent = connectToProps(Component, actions);

			const StateContainer = createStateContainer(BoundComponent);

			expect(() => {
				renderer.create(
					<StateContainer></StateContainer>
				);
			}).toThrowErrorMatchingSnapshot();

		});

		describe('promised state', () => {

			it('should resolve promise state', () => {

				const actions = {
					action1: () => {
						return Promise.resolve({ resolvedProp1: 'resolvedProp' });
					},
				};

				const Component = (props) => {
					return (
						<div {...props}></div>
					);
				};

				Component.propTypes = {
					resolvedProp1: React.PropTypes.string,
				};

				const BoundComponent = connectToProps(Component, actions);

				const StateContainer = createStateContainer(BoundComponent);

				const component = renderer.create(
					<StateContainer></StateContainer>
				);

				let tree = component.toJSON();
				let props = tree.props;

				return props.action1().then(() => {
					tree = component.toJSON();
					props = tree.props;
					return expect(_.omit(props, 'action1')).toEqual({ resolvedProp1: 'resolvedProp' });
				});

			});

			it('should resolve only last action call', () => {

				function returnFirst() {
					return new Promise((resolve) => {
						setTimeout(() => {
							resolve({ resolvedProp1: 1 });
						}, 10);
					});
				}

				function returnSecond() {
					return new Promise((resolve) => {
						setTimeout(() => {
							resolve({ resolvedProp1: 2 });
						}, 500);
					});
				}

				const functionQueue = [returnFirst, returnSecond];

				const actions = {
					action1: () => {
						return functionQueue.shift()();
					},
				};

				const Component = (props) => {
					return (
						<div {...props}></div>
					);
				};

				Component.propTypes = {
					resolvedProp1: React.PropTypes.number,
				};

				const BoundComponent = connectToProps(Component, actions);

				const StateContainer = createStateContainer(BoundComponent);

				const component = renderer.create(
					<StateContainer></StateContainer>
				);

				let tree = component.toJSON();
				let props = tree.props;

				return Promise.all([
					props.action1(),
					props.action1(),
				]).then(() => {
					tree = component.toJSON();
					props = tree.props;
					return expect(_.omit(props, 'action1')).toEqual({ resolvedProp1: 2 });
				});

			});

		});

	});

	describe('#mockActions', () => {

		it('should mock an action', () => {

			const actions = {
				action1: jest.fn(function(state) {
					return {
						...state,
						value1: 5,
					};
				}),
			};

			const computeds = {};

			const mocked = mockActions(actions, computeds);
			expect(mocked.action1).toBeDefined();

			const result = mocked.action1();
			expect(result.then).toBeDefined();

			return result.then((value) => {
				expect(value).toEqual({ value1: 5 });
				return null;
			});

		});

		it('should default computeds', () => {

			let computeds;

			const actions = {
				action1: jest.fn(function() { computeds = this.getComputed(); }),
			};

			const mocked = mockActions(actions);
			expect(mocked.action1).toBeDefined();

			const result = mocked.action1();
			expect(result.then).toBeDefined();

			return result.then(() => {
				expect(computeds).toEqual({});
				return null;
			});

		});

	});

});
