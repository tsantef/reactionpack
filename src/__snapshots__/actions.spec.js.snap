exports[`Actions wrappedAction should reject invalid action 1`] = `"Actions must be either functions or objects"`;
